ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -1413,7 +1413,7 @@ leaderId, getMemberId(), currentTerm, followerCommit, state.getNextIndex(), NOT_
             getRaftServer().getPeer());
       }
     }
-    return JavaUtils.allOf(futures).whenCompleteAsync(
+    return JavaUtils.allOf(futures).whenComplete(
```

Comment:
> Actions supplied for dependent completions of non-async methods may be performed by the thread that completes the current CompletableFuture, or by any other caller of a completion method.

According to the above javadoc https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html , it uses the thread, which is SegmentedRaftLogWorker in this case, completing `futures` passed to `allOf(..)` but not the thread running `appendEntriesAsync`.   Therefore, we should pass serverExecutor as below
```
    return JavaUtils.allOf(futures).whenCompleteAsync(
        (r, t) -> followerState.ifPresent(fs -> fs.updateLastRpcTime(FollowerState.UpdateType.APPEND_COMPLETE)),
        serverExecutor
    ).thenApply(v -> {
```

Reply:
Thanks @szetszwo for the review.  Passing `serverExecutor` was my first attempt, but it caused timeout in `TestInstallSnapshotNotificationWithGrpc`:

https://github.com/adoroszlai/incubator-ratis/runs/7650650577?check_suite_focus=true#step:5:2830

It may indicate that the test or some other part of Ratis needs further tweak.  I'll take another look.

Reply:
`TestInstallSnapshotNotificationWithGrpc` may be flaky after all, though I haven't found recent commits with the same failure.